src/java_bytecode/java_types.h

```diff
@@ -76,4 +78,124 @@ bool is_java_array_tag(const irep_idt& tag);
 
 bool is_valid_java_array(const struct_typet &type);
 
+
+/// class to hold a Java generic type
+/// upper bound can specify type requirements
+class java_generic_typet:public reference_typet
+{
+public:
+  typet bound;
+  const irep_idt &type_var_name;
+
+  void set_bound(const typet &_bound)
+  {
+    bound=_bound;
+  }
+
+  java_generic_typet(const irep_idt &_type_var_name) :
+    reference_typet(),
+    type_var_name(_type_var_name)
+  {
+    set(ID_java_generic_type, true);
+    // set standard bound
+    set_bound(java_type_from_string("Ljava/lang/Object;"));
+  }
+};
+
+inline bool is_java_generic_type(const typet &type)
+{
+  return type.get_bool(ID_java_generic_type);
+}
+
+inline const java_generic_typet &to_java_generic_type(const typet &type)
+{
+  PRECONDITION(
+    type.id()==ID_pointer &&
+    is_java_generic_type(type));
+  return static_cast<const java_generic_typet &>(type);
+}
+
+inline java_generic_typet &to_java_generic_type(typet &type)
+{
+  PRECONDITION(
+    type.id()==ID_pointer &&
+    is_java_generic_type(type));
+  return static_cast<java_generic_typet &>(type);
+}
+
+/// class to hold instantiated type variable bound is exact, for example the
+/// `java.lang.Integer` type in a `List<Integer>`
+class java_inst_generic_typet:public java_generic_typet
+{
+public:
+  java_inst_generic_typet(const reference_typet &type) :
+    java_generic_typet(irep_idt())
+  {
+    set(ID_java_inst_generic_type, true);
+  }
+};
+
+inline bool is_java_inst_generic_type(const typet &type)
+{
+  return type.get_bool(ID_java_inst_generic_type);
+}
+
+inline const java_inst_generic_typet &to_java_inst_generic_type(const typet &type)
+{
+  PRECONDITION(
+    type.id()==ID_pointer &&
+    is_java_inst_generic_type(type));
+  return static_cast<const java_inst_generic_typet &>(type);
+}
+
+inline java_inst_generic_typet &to_java_inst_generic_type(typet &type)
+{
+  PRECONDITION(
+    type.id()==ID_pointer &&
+    is_java_inst_generic_type(type));
+  return static_cast<java_inst_generic_typet &>(type);
+}
+
+/// class to hold type with generic type variable, for example `java.util.List`
+/// in either List<Integer> or List<T>. The vector holds the types of the type
+/// Variables, that is the vector has the length of the number of type variables
+/// of the generic class. For example in `HashMap<K, V>` it would contain two
+/// elements, each of type `java_generic_typet`, in `HashMap<Integer, V>` it
+/// would contains two elements, the first of type `java_inst_generic_typet` and
+/// the second of type `java_generic_typet`.
+class java_type_with_generic_typet:public reference_typet
```

Comment:
You might consider calling _this_ `java_generic_typet` though if you wish to keep it as this since more explicit then that's fine.